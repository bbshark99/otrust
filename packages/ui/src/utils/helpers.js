export function getFirstMessage(messagesObject) {
  const filteredArray = Object.values(messagesObject).filter(message => message);
  return filteredArray.length ? filteredArray[0] : null;
}
