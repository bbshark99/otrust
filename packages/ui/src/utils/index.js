export const shorter = str => (str?.length > 8 ? `${str.slice(0, 6)}...${str.slice(-4)}` : str);

// export const fetcher = (library: Web3Provider, abi?: any) => (...args) => {
//   const [arg1, arg2, ...params] = args
//   // it's a contract
//   if (isAddress(arg1)) {
//     const address = arg1
//     const method = arg2
//     const contract = new Contract(address, abi, library.getSigner())
//     return contract[method](...params)
//   }
//   // it's a eth call
//   const method = arg1
//   return library[method](arg2, ...params)
// }
