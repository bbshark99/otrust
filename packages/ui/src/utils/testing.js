import React from 'react';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';
import { BondingCurveContext } from '@onomy/react-hub';

import { darkNew } from 'theme/theme';
import { ExchangeContext, UpdateExchangeContext } from '../context/exchange/ExchangeContext';
import { ModalContext } from '../context/modal/ModalContext';

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <MemoryRouter>
        <ThemeProvider theme={darkNew}>
          <Component {...props}>{children}</Component>
        </ThemeProvider>
      </MemoryRouter>
    );
  }
  return render(
    <MemoryRouter>
      <ThemeProvider theme={darkNew}>
        <Component {...props} />
      </ThemeProvider>
    </MemoryRouter>
  );
};

export function ThemeWrapper(Component, props, children) {
  return (
    <ThemeProvider theme={darkNew}>
      <Component {...props}>{children}</Component>
    </ThemeProvider>
  );
}

export function ChainContextWrapper(children, contextProps) {
  return (
    <BondingCurveContext.Provider
      value={{
        blockNumber: BigNumber(0),
        currentETHPrice: BigNumber(0),
        currentNOMPrice: BigNumber(0),
        NOMallowance: BigNumber(0),
        strongBalance: BigNumber(0),
        supplyNOM: BigNumber(0),
        weakBalance: BigNumber(0),
        ...contextProps,
      }}
    >
      {children}
    </BondingCurveContext.Provider>
  );
}

export function ExchangeContextWrapper(children, contextProps) {
  return (
    <ExchangeContext.Provider
      value={{
        askAmount: BigNumber(0),
        bidAmount: BigNumber(0),
        bidDenom: 'strong',
        status: 'Not Approved',
        strong: 'ETH',
        weak: 'NOM',
        ...contextProps,
      }}
    >
      {children}
    </ExchangeContext.Provider>
  );
}

export function UpdateExchangeContextWrapper(children, contextProps) {
  return (
    <UpdateExchangeContext.Provider
      value={{
        objDispatch: () => {},
        strDispatch: () => {},
        setInputPending: false,
        ...contextProps,
      }}
    >
      {children}
    </UpdateExchangeContext.Provider>
  );
}

export function ModalContextWrapper(children, contextProps) {
  return (
    <ModalContext.Provider
      value={{
        handleModal: () => {},
        modal: false,
        modalContent: 'Modal Content',
        ...contextProps,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
}

export const renderWithContext = (Component, props, contextValues) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <BondingCurveContext.Provider
        value={{
          supplyNOM: BigNumber(0),
          blockNumber: BigNumber(0),
          currentETHPrice: BigNumber(0),
          currentNOMPrice: BigNumber(0),
          NOMallowance: BigNumber(0),
          strongBalance: BigNumber(0),
          weakBalance: BigNumber(0),
          ...contextValues,
        }}
      >
        <UpdateExchangeContext.Provider
          value={{
            objDispatch: jest.fn(),
            strDispatch: jest.fn(),
            setInputPending: false,
            ...contextValues,
          }}
        >
          <ExchangeContext.Provider
            value={{
              askAmount: BigNumber(0),
              bidAmount: BigNumber(0),
              bidDenom: 'strong',
              status: 'Not Approved',
              strong: 'ETH',
              weak: 'NOM',
              ...contextValues,
            }}
          >
            <ModalContext.Provider
              value={{
                handleModal: jest.fn(),
                modal: false,
                modalContent: 'Modal Content',
                ...contextValues,
              }}
            >
              <Component {...props} />
            </ModalContext.Provider>
          </ExchangeContext.Provider>
        </UpdateExchangeContext.Provider>
      </BondingCurveContext.Provider>
    </ThemeProvider>
  );
};
