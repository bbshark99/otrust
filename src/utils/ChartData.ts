import { ApolloClient, NormalizedCacheObject, gql } from '@apollo/client';
import { BigNumber } from 'bignumber.js';

import {
  ErrorCallback,
  HistoryCallback,
  IDatafeedChartApi,
  LibrarySymbolInfo,
  OnReadyCallback,
  PeriodParams,
  ResolutionString,
  ResolveCallback,
  SearchSymbolsCallback,
  SubscribeBarsCallback,
} from 'charting_library/charting_library';

const WNOM_HISTORICAL_DATA_QUERY = gql`
  query transactions($filter: String!, $skipped: Int!) {
    wnomhistoricalFrames(
      first: 1000
      skip: $skipped
      where: { type: $filter }
      orderBy: startTime
      orderDirection: asc
    ) {
      startTime
      endTime
      startPrice
      endPrice
      minPrice
      maxPrice
    }
  }
`;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const PERIOD_MAP = {
  '1': 'Minute',
  '15': 'QuarterHour',
  '60': 'Hour',
  '480': 'QuarterDay',
  '1D': 'Day',
  '1W': 'Week',
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const NOM_SYMBOL_INFO: LibrarySymbolInfo = {
  name: 'NOM',
  full_name: 'NOM',
  description: 'NOM',
  exchange: 'EthBondCurve',
  listed_exchange: 'EthBondCurve',
  format: 'price',
  has_intraday: true,
  has_no_volume: true,
  has_seconds: false,
  has_ticks: false,
  has_daily: true,
  has_empty_bars: true,
  data_status: 'streaming',
  original_currency_code: 'ETH',
  currency_code: 'ETH',
  // @ts-ignore
  supported_resolutions: Object.keys(PERIOD_MAP),
  intraday_multipliers: Object.keys(PERIOD_MAP),
  //supported_resolutions: [],
  ticker: 'NOM',
  timezone: 'Etc/UTC',
  type: 'crypto',
  session: '24x7',
  minmov: 1,
  pricescale: 10000000000,
};

function formatPrice(price: string) {
  return new BigNumber(price).shiftedBy(-18).toFixed(10);
}

// @ts-ignore
export class ChartData implements IDatafeedChartApi {
  private apolloClient: ApolloClient<NormalizedCacheObject>;

  constructor(apolloClient: ApolloClient<NormalizedCacheObject>) {
    this.apolloClient = apolloClient;
  }

  public async onReady(cb: OnReadyCallback) {
    cb({
      exchanges: [
        {
          value: NOM_SYMBOL_INFO.exchange,
          name: NOM_SYMBOL_INFO.exchange,
          desc: NOM_SYMBOL_INFO.exchange,
        },
      ],
      symbols_types: [
        {
          value: NOM_SYMBOL_INFO.type,
          name: NOM_SYMBOL_INFO.type,
        },
      ],
      supported_resolutions: NOM_SYMBOL_INFO.supported_resolutions,
      currency_codes: ['ETH'],
      units: {},
      supports_marks: false,
      supports_timescale_marks: false,
      supports_time: true,
    });
  }

  public async searchSymbols(
    userInput: string,
    exchange: string,
    symbolType: string,
    onResult: SearchSymbolsCallback
  ) {
    onResult([
      {
        symbol: NOM_SYMBOL_INFO.ticker!,
        full_name: NOM_SYMBOL_INFO.name,
        description: NOM_SYMBOL_INFO.description,
        exchange: NOM_SYMBOL_INFO.exchange,
        ticker: NOM_SYMBOL_INFO.ticker!,
        type: NOM_SYMBOL_INFO.type,
      },
    ]);
  }

  resolveSymbol(symbolName: string, onSuccess: ResolveCallback) {
    // TradingView requires this setTimeout
    setTimeout(() => onSuccess(NOM_SYMBOL_INFO), 0);
  }

  public async getBars(
    symbolInfo: LibrarySymbolInfo,
    resolution: string,
    periodParams: PeriodParams,
    onResult: HistoryCallback,
    onError: ErrorCallback
  ) {
    try {
      const resp = await this.apolloClient.query({
        query: WNOM_HISTORICAL_DATA_QUERY,
        variables: {
          // @ts-ignore
          filter: PERIOD_MAP[resolution] || 'Hour',
          skipped: 0,
        },
      });

      const {
        data: { wnomhistoricalFrames: frames },
      } = resp;

      const bars = frames.map(
        (frame: {
          startTime: any;
          startPrice: any;
          maxPrice: any;
          minPrice: any;
          endPrice: any;
        }) => ({
          time: parseInt(frame.startTime, 10) * 1000,
          open: formatPrice(frame.startPrice),
          high: formatPrice(frame.maxPrice),
          low: formatPrice(frame.minPrice),
          close: formatPrice(frame.endPrice),
          volume: 0,
        })
      );

      console.log('bars', bars);

      if (periodParams.firstDataRequest) {
        onResult(bars, { noData: false });
      } else {
        onResult([], { noData: true });
      }
    } catch (e) {
      onError(`${e}`);
    }
  }

  public async subscribeBars(
    symbolInfo: LibrarySymbolInfo,
    resolution: ResolutionString,
    onTick: SubscribeBarsCallback,
    listenerGuid: string,
    onResetCacheNeededCallback: () => void
  ) {
    console.log('subscribeBars', {
      symbolInfo,
      resolution,
      listenerGuid,
      onResetCacheNeededCallback,
    });

    // TODO: implement data subscription
  }

  public async unsubscribeBars(listenerGuid: string) {
    console.log('unsubscribeBars', listenerGuid);
    // TODO: implement data unsubscribe
  }
}
