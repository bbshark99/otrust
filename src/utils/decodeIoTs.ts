import { isLeft, getOrElse } from 'fp-ts/lib/Either';
import * as t from 'io-ts';

export function decodeIoTs<T, O, I>(validator: t.Type<T, O, I>, input: I): T {
  const result = validator.decode(input);
  const decoded = getOrElse(() => result as unknown as T)(result);

  if (isLeft(result)) {
    throw new Error(result.left.toString());
  }

  return decoded;
}

export function encodeIoTs<T, O, I>(validator: t.Type<T, O, I>, input: T): O {
  const result = validator.encode(input);
  return result;
}
