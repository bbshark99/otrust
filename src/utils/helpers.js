export function getFirstMessage(messagesObject) {
  let filteredArray = Object.values(messagesObject).filter(message => message);
  return filteredArray.length ? filteredArray[0] : null;
}
