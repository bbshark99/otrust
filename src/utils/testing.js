import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';
import { ModalContext } from '../context/modal/ModalContext';
import { UpdateExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const ThemeWrapper = (Component, props, children) => {
  return (
    <ThemeProvider theme={darkNew}>
      <Component {...props}>{children}</Component>
    </ThemeProvider>
  );
};

export const ChainContextWrapper = (children, contextProps) => {
  return (
    <ChainContext.Provider
      value={{
        blockNumber: BigNumber(0),
        currentETHPrice: BigNumber(0),
        currentNOMPrice: BigNumber(0),
        NOMallowance: BigNumber(0),
        strongBalance: BigNumber(0),
        supplyNOM: BigNumber(0),
        weakBalance: BigNumber(0),
        theme: { darkNew },
        ...contextProps,
      }}
    >
      {children}
    </ChainContext.Provider>
  );
};

export const ExchangeContextWrapper = (children, contextProps) => {
  return (
    <ExchangeContext.Provider
      value={{
        askAmount: BigNumber(0),
        bidAmount: BigNumber(0),
        bidDenom: 'strong',
        status: 'Not Approved',
        strong: 'ETH',
        weak: 'NOM',
        ...contextProps,
      }}
    >
      {children}
    </ExchangeContext.Provider>
  );
};

export const UpdateExchangeContextWrapper = (children, contextProps) => {
  return (
    <UpdateExchangeContext.Provider
      value={{ objDispatch: () => {}, strDispatch: () => {}, setInputPending: false, ...contextProps }}
    >
      {children}
    </UpdateExchangeContext.Provider>
  );
};

export const ModalContextWrapper = (children, contextProps) => {
  return (
    <ModalContext.Provider
      value={{ handleModal: () => {}, modal: false, modalContent: 'Modal Content', ...contextProps }}
    >
      {children}
    </ModalContext.Provider>
  );
};

export const renderWithContext = (Component, props, contextValues) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider
          value={{
            supplyNOM: BigNumber(0),
            blockNumber: BigNumber(0),
            currentETHPrice: BigNumber(0),
            currentNOMPrice: BigNumber(0),
            NOMallowance: BigNumber(0),
            strongBalance: BigNumber(0),
            weakBalance: BigNumber(0),
            theme: darkNew,
            ...contextValues,
          }}
        >
          <UpdateExchangeContext.Provider
            value={{ objDispatch: jest.fn(), strDispatch: jest.fn(), setInputPending: false, ...contextValues }}
          >
            <ExchangeContext.Provider
              value={{
                askAmount: BigNumber(0),
                bidAmount: BigNumber(0),
                bidDenom: 'strong',
                status: 'Not Approved',
                strong: 'ETH',
                weak: 'NOM',
                ...contextValues,
              }}
            >
              <ModalContext.Provider
                value={{ handleModal: jest.fn(), modal: false, modalContent: 'Modal Content', ...contextValues }}
              >
                <Component {...props} />
              </ModalContext.Provider>
            </ExchangeContext.Provider>
          </UpdateExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
