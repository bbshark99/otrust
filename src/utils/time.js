const SECONDS_IN_ONE_DAY = 86400;

export const getTimeStampDaysAgo = numberOfDays => {
  const date = new Date();
  return Math.floor(date.getTime() / 1000 - numberOfDays * SECONDS_IN_ONE_DAY).toString();
};

export const getTimeStampMonthsAgo = numberOfMonths => {
  const date = new Date();
  date.setMonth(date.getMonth() - numberOfMonths);
  return Math.floor(date.getTime() / 1000).toString();
};
