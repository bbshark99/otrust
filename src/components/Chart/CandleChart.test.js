import { cleanup } from '@testing-library/react';

import CandleChart from './CandleChart';
import { renderWithContext } from '../../utils/testing';

describe('Given the CandleChart component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithContext(CandleChart);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
