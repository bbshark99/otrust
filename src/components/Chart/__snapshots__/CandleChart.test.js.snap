// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the CandleChart component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="candle-chart"
    style="margin-top: 1rem; height: 350px;"
  >
    <svg
      class="sc-bdnxRM kNdflx"
      data-testid="candle-chart-svg"
    >
      <g
        class="y-axis"
      />
      <g
        class="x-axis1"
      />
      <g
        class="x-axis"
      />
      <g
        class="candlestickGroup"
      />
    </svg>
  </div>
</DocumentFragment>
`;
